bittorrent/sanitize.go

```diff
@@ -0,0 +1,50 @@
+package bittorrent
+
+import (
+	"errors"
+	"net"
+)
+
+// ErrInvalidIP indicates an invalid IP for an Announce.
+var ErrInvalidIP = errors.New("invalid IP")
+
+// RequestSanitizer is used to replace unreasonable values in requests parsed
+// from a frontend into sane values.
+type RequestSanitizer struct {
+	MaxNumWant          uint32 `yaml:"max_numwant"`
+	DefaultNumWant      uint32 `yaml:"default_numwant"`
+	MaxScrapeInfoHashes uint32 `yaml:"max_scrape_infohashes"`
+}
+
+// SanitizeAnnounce enforces a max and default NumWant and coerces the a peer's
+// IP address into the proper format.
+func (rs *RequestSanitizer) SanitizeAnnounce(r *AnnounceRequest) (*AnnounceRequest, error) {
+	if r.NumWant > rs.MaxNumWant {
+		r.NumWant = rs.MaxNumWant
+	}
+
+	if r.NumWant == 0 {
```

Comment:
Hmm, we have to do something like that. But I want to avoid allocating a whole new struct for it.

Maybe something like:

```go
type AnnounceRequest struct {
    NumWant OptionalNumWant
    ...
}

type OptionalNumWant struct {
    Value uint32
    Provided bool
}
```

Then sanitize accordingly and guarantee that `NumWant.Value` is set to a sane value?

Or, if we want to be really high performance and dirty, just
```go
const NumWantNotSet uint32(13337)
```
and check against that :)

I'm thinking - what are the security implications if someone can trick us, with a UDP announce, into believing no NumWant was specified? We're gonna use our default and they'll know our default, but who really cares?